pkg/wasm/convert.go

```diff
@@ -0,0 +1,153 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wasm
+
+import (
+	"sync"
+	"time"
+
+	udpa "github.com/cncf/udpa/go/udpa/type/v1"
+	core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	wasm "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/wasm/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/conversion"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/any"
+	"go.uber.org/atomic"
+)
+
+const (
+	apiTypePrefix      = "type.googleapis.com/"
+	typedStructType    = apiTypePrefix + "udpa.type.v1.TypedStruct"
+	wasmHTTPFilterType = apiTypePrefix + "envoy.extensions.filters.http.wasm.v3.Wasm"
+)
+
+// MaybeConvertWasmExtensionConfig converts any presence of module remote download to local file.
+// It downloads the Wasm module and stores the module locally in the file system.
+func MaybeConvertWasmExtensionConfig(resources []*any.Any, cache Cache) bool {
+	var wg sync.WaitGroup
+	numResources := len(resources)
+	wg.Add(numResources)
+	sendNack := atomic.NewBool(false)
+
+	for i := 0; i < numResources; i++ {
+		go func(i int) {
+			defer wg.Done()
+
+			newExtensionConfig, nack := convert(resources[i], cache)
+			if nack && !sendNack.Load() {
```

Comment:
why do we need `&& !sendNack.Load()`? this means if we have 2 failures, the latter will still update `resources` I think?

Reply:
Ah, yeah, we don't need to check here. fixed.